# Data Extraction from Relational Database (SQLite)

In [1]:
# libraries
import sqlite3
import pandas as pd
import prettytable
from prettytable import TableStyle

In [2]:
prettytable.DEFAULT = TableStyle.DEFAULT

In [3]:
# load extension
%load_ext sql

# establish connection for the SQL magic module to the sql server
%sql sqlite:///../nepal.sqlite

## Check number of tables in database

In [4]:
%%sql
SELECT name FROM sqlite_schema
WHERE type = "table";

 * sqlite:///../nepal.sqlite
Done.


[('id_map',),
 ('building_structure',),
 ('building_damage',),
 ('household_demographics',)]

## Inspect individual table

### id_map

In [5]:
%%sql

SELECT * FROM id_map
LIMIT 5;

 * sqlite:///../nepal.sqlite
Done.


[(5601, 56, 7, 1),
 (6301, 63, 7, 1),
 (9701, 97, 7, 1),
 (9901, 99, 7, 1),
 (11501, 115, 7, 1)]

In [6]:
%%sql

SELECT count(*) as number_of_records 
FROM id_map;

 * sqlite:///../nepal.sqlite
Done.


[(249932,)]

In [7]:
%%sql

SELECT count(DISTINCT building_id ) as num_building
FROM id_map;

 * sqlite:///../nepal.sqlite
Done.


[(234835,)]

In [8]:
%%sql

SELECT count(DISTINCT district_id ) as num_district
FROM id_map;

 * sqlite:///../nepal.sqlite
Done.


[(4,)]

### building_structure

In [9]:
%%sql

SELECT * FROm building_structure
LIMIT 5;

 * sqlite:///../nepal.sqlite
Done.


[(1, 1, 1, 9, 288, 9, 9, 'Flat', 'Other', 'Bamboo/Timber-Light roof', 'Mud', 'Not applicable', 'Not attached', 'Rectangular', 'Damaged-Used in risk', 'Stone, mud mortar'),
 (2, 1, 1, 15, 364, 9, 9, 'Flat', 'Other', 'Bamboo/Timber-Light roof', 'Mud', 'Not applicable', 'Not attached', 'Rectangular', 'Damaged-Repaired and used', 'Stone, mud mortar'),
 (3, 1, 1, 20, 384, 9, 9, 'Flat', 'Other', 'Bamboo/Timber-Light roof', 'Mud', 'Not applicable', 'Not attached', 'Rectangular', 'Damaged-Repaired and used', 'Stone, mud mortar'),
 (4, 1, 1, 20, 312, 9, 9, 'Flat', 'Other', 'Bamboo/Timber-Light roof', 'Mud', 'Not applicable', 'Not attached', 'Rectangular', 'Damaged-Repaired and used', 'Stone, mud mortar'),
 (5, 1, 1, 30, 308, 9, 9, 'Flat', 'Other', 'Bamboo/Timber-Light roof', 'Mud', 'Not applicable', 'Not attached', 'Rectangular', 'Damaged-Repaired and used', 'Stone, mud mortar')]

In [10]:
%%sql

SELECT count(DISTINCT building_id ) as num_building
FROM building_structure;

 * sqlite:///../nepal.sqlite
Done.


[(234835,)]

### building_damage

In [11]:
%%sql
SELECT * FROM building_damage
LIMIT 5;

 * sqlite:///../nepal.sqlite
Done.


[(1, 'Moderate-Heavy', 'Insignificant/light', None, None, 'Moderate-Heavy-(<1/3)', 'Insignificant/light-(<1/3)', 'Severe-Extreme-(<1/3)', None, 'Insignificant/light-(>2/3)', 'Severe-Extreme-(>2/3)', None, None, 'Severe-Extreme-(<1/3)', None, None, 'Severe-Extreme-(>2/3)', None, None, None, 'Moderate-Heavy-(>2/3)', None, 'Severe-Extreme-(>2/3)', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'Both', 'Grade 3', 'Major repair', 0.0, None, 1.0, 1.0, 1.0, 1.0, None, 1.0, 0.0, 0.0, None, None, None, 0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0, 0, 0),
 (2, 'Severe-Extreme', 'Severe-Extreme', 'Insignificant/light', 'Severe-Extreme-(>2/3)', None, None, 'Severe-Extreme-(>2/3)', None, None, 'Severe-Extreme-(>2/3)', None, None, 'Severe-Extreme-(>2/3)', None, None, 'Severe-Extreme-(>2/3)', None, None, 'Severe-Extreme-(>2/3)', None, None, 'Severe-Extreme-(>2/3)', None, None, None, None, None, 'Severe-Extreme-(>2/3)', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'Exterior', 'Grade 5', 'Reconstruction', 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 1.0, None, None, None, 0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0, 0, 0),
 (3, 'Moderate-Heavy', 'Moderate-Heavy', 'Moderate-Heavy', None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, None, 'Insignificant/light-(1/3-2/3)', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'Both', 'Grade 2', 'Minor repair', 1.0, None, None, None, None, None, None, None, None, None, None, None, None, 1.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0, 0, 0),
 (4, 'Moderate-Heavy', 'Moderate-Heavy', 'Moderate-Heavy', None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, 'Moderate-Heavy-(>2/3)', None, None, None, 'Insignificant/light-(<1/3)', None, None, 'Insignificant/light-(1/3-2/3)', None, None, None, None, None, None, None, None, None, None, 'Moderate-Heavy-(>2/3)', None, None, None, None, None, None, None, 'Both', 'Grade 2', 'Minor repair', 1.0, None, None, None, None, None, None, None, None, None, None, None, None, None, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0, 0, 0),
 (5, 'Insignificant/light', None, None, None, None, 'Insignificant/light-(<1/3)', None, None, 'Insignificant/light-(<1/3)', None, None, 'Insignificant/light-(<1/3)', None, None, 'Insignificant/light-(<1/3)', None, None, 'Insignificant/light-(<1/3)', None, None, 'Insignificant/light-(<1/3)', None, None, 'Insignificant/light-(<1/3)', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'Exterior', 'Grade 1', 'Minor repair', 1.0, None, None, None, None, None, None, None, 0.0, 0.0, None, None, None, 0.0, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0, 0, 0)]

In [12]:
%%sql

SELECT count(distinct building_id) as num_building
FROM building_damage;

 * sqlite:///../nepal.sqlite
Done.


[(234835,)]

In [13]:
%%sql

SELECT damage_grade, technical_solution_proposed
FROM building_damage
GROUP BY damage_grade;

 * sqlite:///../nepal.sqlite
Done.


[(None, None),
 ('Grade 1', 'Minor repair'),
 ('Grade 2', 'Minor repair'),
 ('Grade 3', 'Major repair'),
 ('Grade 4', 'Major repair'),
 ('Grade 5', 'Reconstruction')]

### household_demographics

In [14]:
%%sql

SELECT * FROM household_demographics
LIMIT 5;

 * sqlite:///../nepal.sqlite
Done.


[(101, 'Male', 31.0, 'Rai', 'Illiterate', 'Rs. 10 thousand', 3.0, 0.0),
 (201, 'Female', 62.0, 'Rai', 'Illiterate', 'Rs. 10 thousand', 6.0, 0.0),
 (301, 'Male', 51.0, 'Gharti/Bhujel', 'Illiterate', 'Rs. 10 thousand', 13.0, 0.0),
 (401, 'Male', 48.0, 'Gharti/Bhujel', 'Illiterate', 'Rs. 10 thousand', 5.0, 0.0),
 (501, 'Male', 70.0, 'Gharti/Bhujel', 'Illiterate', 'Rs. 10 thousand', 8.0, 0.0)]

From the inspection data, the surveyor assessed 234,835 buildings across 4 districts. In the building_damage table, the damage_grade column provides a concise summary of the extent of damage each building sustained from the earthquake. The surveyor also recommended a technical solution for each building: buildings with damage grades 3 and 4 were proposed for major repairs, while those with grade 5 were slated for reconstruction. Based on this, damage grades 3, 4, and 5 can be grouped as severely damaged buildings, making the damage_grade column a suitable target variable for analysis.

## Extract table for Analysis and Modelling

In [15]:
%%sql
 
SELECT distinct i.building_id as b_id, i.district_id, s.*, d.damage_grade
FROM id_map as i
JOIN building_structure as s ON i.building_id = s.building_id
JOIN building_damage as d ON i.building_id = d.building_id
LIMIT 5;

 * sqlite:///../nepal.sqlite
Done.


[(56, 1, 56, 2, 2, 40, 322, 18, 18, 'Flat', 'Mud mortar-Stone/Brick', 'Bamboo/Timber-Light roof', 'Mud', 'TImber/Bamboo-Mud', 'Not attached', 'Rectangular', 'Damaged-Not used', 'Stone, mud mortar', 'Grade 2'),
 (63, 1, 63, 2, 2, 1, 437, 16, 16, 'Flat', 'Mud mortar-Stone/Brick', 'Bamboo/Timber-Light roof', 'Mud', 'Timber-Planck', 'Not attached', 'Rectangular', 'Not damaged', 'Stone, mud mortar', 'Grade 1'),
 (97, 1, 97, 2, 2, 22, 420, 16, 16, 'Flat', 'Mud mortar-Stone/Brick', 'Bamboo/Timber-Light roof', 'Mud', 'TImber/Bamboo-Mud', 'Not attached', 'Rectangular', 'Damaged-Not used', 'Stone, mud mortar', 'Grade 2'),
 (99, 1, 99, 2, 2, 50, 242, 16, 16, 'Flat', 'Mud mortar-Stone/Brick', 'Bamboo/Timber-Light roof', 'Mud', 'TImber/Bamboo-Mud', 'Not attached', 'Rectangular', 'Damaged-Not used', 'Stone, mud mortar', 'Grade 4'),
 (115, 1, 115, 2, 2, 12, 308, 16, 16, 'Flat', 'Mud mortar-Stone/Brick', 'Bamboo/Timber-Light roof', 'Mud', 'Timber-Planck', 'Not attached', 'Rectangular', 'Damaged-Not used', 'Stone, mud mortar', 'Grade 3')]

In [16]:
# create a connection to the database
conn = sqlite3.connect("../nepal.sqlite")

In [19]:
# Read in data into pandas dataframe
query = """
SELECT distinct i.building_id as b_id, i.district_id, s.*, d.damage_grade
FROM id_map as i
JOIN building_structure as s ON i.building_id = s.building_id
JOIN building_damage as d ON i.building_id = d.building_id
"""

df = pd.read_sql(query, conn)
df.head()

b_id  district_id  building_id  count_floors_pre_eq  count_floors_post_eq  \
0    56            1           56                    2                     2   
1    63            1           63                    2                     2   
2    97            1           97                    2                     2   
3    99            1           99                    2                     2   
4   115            1          115                    2                     2   

   age_building  plinth_area_sq_ft  height_ft_pre_eq  height_ft_post_eq  \
0            40                322                18                 18   
1             1                437                16                 16   
2            22                420                16                 16   
3            50                242                16                 16   
4            12                308                16                 16   

  land_surface_condition         foundation_type                 roof_type  \
0                   Flat  Mud mortar-Stone/Brick  Bamboo/Timber-Light roof   
1                   Flat  Mud mortar-Stone/Brick  Bamboo/Timber-Light roof   
2                   Flat  Mud mortar-Stone/Brick  Bamboo/Timber-Light roof   
3                   Flat  Mud mortar-Stone/Brick  Bamboo/Timber-Light roof   
4                   Flat  Mud mortar-Stone/Brick  Bamboo/Timber-Light roof   

  ground_floor_type   other_floor_type      position plan_configuration  \
0               Mud  TImber/Bamboo-Mud  Not attached        Rectangular   
1               Mud      Timber-Planck  Not attached        Rectangular   
2               Mud  TImber/Bamboo-Mud  Not attached        Rectangular   
3               Mud  TImber/Bamboo-Mud  Not attached        Rectangular   
4               Mud      Timber-Planck  Not attached        Rectangular   

  condition_post_eq     superstructure damage_grade  
0  Damaged-Not used  Stone, mud mortar      Grade 2  
1       Not damaged  Stone, mud mortar      Grade 1  
2  Damaged-Not used  Stone, mud mortar      Grade 2  
3  Damaged-Not used  Stone, mud mortar      Grade 4  
4  Damaged-Not used  Stone, mud mortar      Grade 3

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 234835 entries, 0 to 234834
Data columns (total 19 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   b_id                    234835 non-null  int64 
 1   district_id             234835 non-null  int64 
 2   building_id             234835 non-null  int64 
 3   count_floors_pre_eq     234835 non-null  int64 
 4   count_floors_post_eq    234835 non-null  int64 
 5   age_building            234835 non-null  int64 
 6   plinth_area_sq_ft       234835 non-null  int64 
 7   height_ft_pre_eq        234835 non-null  int64 
 8   height_ft_post_eq       234835 non-null  int64 
 9   land_surface_condition  234835 non-null  object
 10  foundation_type         234835 non-null  object
 11  roof_type               234835 non-null  object
 12  ground_floor_type       234835 non-null  object
 13  other_floor_type        234835 non-null  object
 14  position                234834 non-n

In [21]:
df.to_csv("../data/nepal_earthquake_damage.csv", index=False)